src/control/ras/event.go

```diff
@@ -0,0 +1,205 @@
+	EventTypeInfoOnly    EventTypeID = C.RAS_TYPE_INFO_ONLY
+)
+
+func (typ EventTypeID) String() string {
+	return C.GoString(C.ras_event_type_enum_to_name(uint32(typ)))
+}
+
+// Uint32 returns uint32 representation of event type.
+func (typ EventTypeID) Uint32() uint32 {
+	return uint32(typ)
+}
+
+// Event describes details of a specific RAS event.
+type Event struct {
+	Name        string `json:"name"`
+	Timestamp   string `json:"timestamp"`
+	Msg         string `json:"msg"`
+	Hostname    string `json:"hostname"`
+	Data        []byte `json:"data"`
+	Rank        uint32 `json:"rank"`
+	InstanceIdx uint32 `json:"instance_idx"`
+	ID          EventID
+	Severity    EventSeverityID
+	Type        EventTypeID
+}
+
+// MarshalJSON marshals ras.Event to JSON.
+func (evt *Event) MarshalJSON() ([]byte, error) {
+	// use a type alias to leverage the default marshal for
+	// most fields
+	type toJSON Event
+	return json.Marshal(&struct {
+		ID       uint32
+		Severity uint32
+		Type     uint32
+		*toJSON
+	}{
+		ID:       evt.ID.Uint32(),
+		Severity: evt.Severity.Uint32(),
+		Type:     evt.Type.Uint32(),
+		toJSON:   (*toJSON)(evt),
+	})
+}
+
+// UnmarshalJSON unmarshals ras.Event from JSON.
+func (evt *Event) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" {
+		return nil
+	}
+
+	// use a type alias to leverage the default unmarshal for
+	// most fields
+	type fromJSON Event
+	from := &struct {
+		ID       uint32
+		Severity uint32
+		Type     uint32
+		*fromJSON
+	}{
+		fromJSON: (*fromJSON)(evt),
+	}
+
+	if err := json.Unmarshal(data, from); err != nil {
+		return err
+	}
+
+	evt.ID = EventID(from.ID)
+	evt.Severity = EventSeverityID(from.Severity)
+	evt.Type = EventTypeID(from.Type)
+
+	return nil
+}
+
+// NewRankFailEvent creates a specific RAS event entry.
+//
+// Hostname should be populated by caller.
+func NewRankFailEvent(instanceIdx uint32, rank uint32, exitErr error) *Event {
+	evt := &Event{
+		Name:        EventRankFail.String(),
+		Timestamp:   common.FormatTime(time.Now().UTC()),
+		Msg:         EventRankFail.Desc(),
+		InstanceIdx: instanceIdx,
+		ID:          EventRankFail,
+		Rank:        rank,
+		Type:        EventTypeStateChange,
+		Severity:    EventSeverityInfo,
+	}
+
+	if exitErr != nil {
+		evt.Severity = EventSeverityError
+		// encode exit error message in event
+		// marshal on string will never fail
+		errBytes, _ := json.Marshal(exitErr.Error())
+		evt.Data = errBytes
+	}
+
+	return evt
+}
+
+// ProcessEvent evaluates and actions the given RAS event.
+func ProcessEvent(log logging.Logger, evt *Event) error {
+	if evt == nil {
+		return errors.New("attempt to process nil ras event")
+	}
+	log.Debugf("processing RAS event %q from rank %d on host %q", evt.Msg, evt.Rank,
+		evt.Hostname)
+
+	switch evt.ID {
+	case EventRankFail, EventRankNoResp:
+		log.Debugf("processing %s (%s) event", evt.ID.String(), evt.ID.Desc())
+	default:
+		return errors.Errorf("unrecognised event ID: %d", evt.ID)
+	}
+
+	return nil
+}
```

Comment:
I'm not sure this is the right way to go about this. I think what you probably want to do is to have a `ras.Processor` entity that is responsible for delegating incoming events to registered handlers. So e.g. as part of setting up the control plane server we would initialize a `ras.Processor` and then register handlers for various event types. I think that would make this code much easier to test and maintain.

If you wanted to be fancy, you could perhaps expose a channel on the `Processor` which receives events and processes them serially (maybe doing some stream analysis to add throttling/debounce, etc). That channel could then be passed down into areas of the code where there are event sources.

If you didn't want to expose the channel directly, I think you could perhaps instead expose `Process()` and `ProcessSync()` methods which use the channel under the covers. The regular `Process()` method could be asynchronous by starting a new goroutine before sending to the channel (with the assumption that it will eventually get sent) so that the caller isn't blocked on sending. Either way, I think a channel input, single processor goroutine model is going to be best for doing any kind of advanced processing.

Reply:
will look into it